src/main/java/it/pagopa/pm/gateway/utils/VPosResponseUtils.java

```diff
@@ -0,0 +1,229 @@
+
+    private String getValue(Element element, VPosResponseEnum authResponseEnum) {
+        return element.getChildText(authResponseEnum.getTagName());
+    }
+
+    public void validateResponseMac(String timestamp, String resultCode, String resultMac, StepZeroRequest pgsRequest) {
+        List<String> vposShopData = vPosUtils.getVposShopByIdPsp(pgsRequest.getIdPsp());
+        String configMac = BooleanUtils.isTrue(pgsRequest.getIsFirstPayment()) ?
+                vposShopData.get(MAC_FIRST_PAY_POSITION) :
+                vposShopData.get(MAC_NEXT_PAY_POSITION);
+        String mac = calculateMac(timestamp, resultCode, configMac);
+        if (isMacDifferent(mac, resultMac)) {
+            String macWithoutShopKey = calculateMac(timestamp, resultCode, NULL_STRING);
+            if (isMacDifferent(macWithoutShopKey, resultMac)) {
+                log.warn(String.format(WRONG_MAC_MSG, macWithoutShopKey, resultMac));
+            }
+        }
+    }
+
+    private String calculateMac(String timestamp, String resultCode, String mac) {
+        VPosMacBuilder vposMacBuilder = new VPosMacBuilder();
+        vposMacBuilder.addString(timestamp);
+        vposMacBuilder.addString(resultCode);
+        vposMacBuilder.addString(mac);
+        return vposMacBuilder.toSha1Hex(DEFAULT_CHARSET);
+    }
+
+    private boolean isMacDifferent(String mac, String responseMac) {
+        return StringUtils.isNotBlank(responseMac) || !StringUtils.equalsIgnoreCase(responseMac, mac);
+    }
+
+    public AuthResponse buildAuthResponse(byte[] clientResponse) throws IOException {
+        AuthResponse authResponse;
+        try {
+            SAXBuilder saxBuilder = new SAXBuilder();
+            saxBuilder.setFeature(DISALLOW_DOCTYPE_DECL, true);
+            saxBuilder.setFeature(EXTERNAL_GENERAL_ENTITIES, false);
+            saxBuilder.setFeature(EXTERNAL_PARAMETER_ENTITIES, false);
+            Document responseDocument = saxBuilder.build(new ByteArrayInputStream(clientResponse));
+            Element root = responseDocument.getRootElement();
+            authResponse = new AuthResponse();
+            authResponse.setTimestamp(getValue(root, TIMESTAMP_RESPONSE));
+            authResponse.setResultCode(getValue(root, RESULT));
+            authResponse.setResultMac(getValue(root, MAC_RESPONSE));
+            Element data = root.getChild(DATA_RESPONSE.getTagName());
+            if (data != null) {
+                Element operation = data.getChild(OPERATION.getTagName());
+                if (operation != null) {
+                    Element authorization = operation.getChild(AUTHORIZATION.getTagName());
+                    checkAuthorization(authResponse, authorization);
+                }
+            }
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+            throw new IOException(e);
+        }
+        return authResponse;
+    }
+
+    private void checkAuthorization(AuthResponse authResponse, Element authorization) {
```

Comment:
Is it ok if these fields are not added to the authorization response?

Reply:
I think so, it is implemented like this on the PM